### ISSUE TITLE
2025 Priorities

### ISSUE THREAD

**Issue by rickeylev** [bazel-contrib/rules_python#2520]
This issue is to capture some of the topics that we, the rules_python maintainers, are particularly interested in this coming year. These are topics where we'd be particularly excited by PRs or design discussions.

Note that these are not promises of particular work. This is still a community driven project, so people work on what they find interesting or rewarding.

Topics are in no particular order. Discussion for a topic should be in their associated issue. This is not an exclusive list.

- A venv and site-packages layout for dependencies:
    - #2156
- Consolidating `*_bazel.bzl` files, removing `semantics.bzl`, and general cleanup of the core rule implementations:
    - #2522 
- Implementing uv-based toolchain
    - #1975
- Supporting other lockfile formats (e.g. [PEP751](https://peps.python.org/pep-0751/))
    - #2787
- Module-based main (`python -m <name>`)
    - #2539
- Build-phase sdist building
    - #2410
- Solution to implicit dependencies on user-space libraries. Related:
    - #2094
    - #2370
- Implementing a C++ based bootstrap (expanding bootstrap to support e.g. compiled languages)
    - #2500
- Make `bootstrap_impl=script` the default
    - #2521
- Splitting sphinxdocs into its own bzlmod MODULE
    - #2511
- Toolchains for local interpreters
    - #2070
- Stabilize the pulling of platform-specific wheels matching target platform (i.e. `experimental_index_url`):
    - #2363 
    - #2100 
    - #2023
    - #2434
    - #260 
    - #2319
    - #2241
    - #1357 
    - #2559
- Zip stdlib to reduce file count / runfiles overhead
  - #1653 

cc @aignas @groodt

**Comment by aignas** [comment]
I have modified the list above to include more references and I realized that I want to stabilize the #260 work and make it default at some point.

**Comment by aignas** [comment]
@arrdem, the above bullet point summarizes all of the related tickets to close #260 and stabilize the feature.

I think the main issues are two:
1. not having PEP508 support in starlark means that the APIs are brittle and we require to pass a lot of things to `whl_library` as parameters instead of via `load` to the macro phase. The #2692 and then #2629 should address that. Once that is done I can refactor the passing of `experimental_index_platforms` and fix #2319.
2. Then we should rewrite the `experimental_index_url` configuration API to be more similar to `uv.defaults` and `uv.configure`, so that we have the right level of expressions to configure the target platforms. That should also make things harder to misuse.
3. Doing the above will mean that we have a solid foundation to correctly setup dependencies for any python package given the target platforms (very similar what `rules_pycross` is doing) and we can turn on the `experimental_index_url` by default.

Related items but not strictly necessary needed:
1. Separate out the `whl_library` for building with `pip` (ticket TBD), just extracting wheels (once PEP508 is done that would be possible without the interpreter) and setting up sources for `py_sdist` (#2410).
2. *I think there is something else but I can't remember right now*.

That is about it.

**Comment by aignas** [comment]
I think we should have something for 2026. Where do we do this? Do we just adapt the list above or create a new issue? I feel like the general project direction is somewhat still similar to what was written here.

**Comment by rickeylev** [comment]
I created a 2026 priorities issue. Yes, it'll definitely have some overlap with what we have in here.

**Comment by cj81499** [comment]
For convenience: https://github.com/bazel-contrib/rules_python/issues/3570
